### PR TITLE
add a deep learning example supporting training, inference, and model aggregation

### PR DESCRIPTION
**Pull request by mofanv**
### Aim
This example aims to support both trusted training and inference inside the Veracruz runtime. It is currently the server-side application, meaning that 1) for centralized machine learning, all data are provisioned and fed into this application; 2) for federated learning, it conducts the model aggregation (e.g., aggregating all models as one global model).

### Status

1. Model training using `classifier` function (tested on a LeNet model on MNIST dataset)
2. Inference using YOLO model for object detection (tested on Tiny-YOLO on images)
3. Centralized training (Multiple inputs -> one output). I.e., data from multiple providers can be loaded into the example with `data.list` file
4. Support argument parser in two ways: i) `args.cfg` and ii) command-line
5. Support average aggregation for darknet model format weights file (for generic federated learning)
6. `wasmtime` passed
7. ONNX model format aggregation supported
8. Tests of 1) prediction results of training 2) aggregation results
2. Run inside freestanding execution engine
3. ONNX model federated learning: clients use Tensorflow, Pytorch, and then export trained model to ONNX for our aggregation

**Note**: read [README.md](https://github.com/mofanv/veracruz-examples/tree/main/deep-learning-server) for instructions and other details.

### CODE REVIEW

**Comment by mofanv** [comment]
During the test of freestanding EE, it failed when running `fopen()`, a C function compiled to `wasm`. Is this a bug in EE? or I did something wrong with the directory. Any thought? @dominic-mulligan-arm @ShaleXIONG @gbryant-arm 

A small example to reproduce the error is 1) creating a `main.c`:
```
#include <stdio.h>
#include <stdlib.h>
int main()
{
   char ch;
   FILE *fp;

   fp = fopen("/test.txt", "r");
   if (fp == NULL){
      perror("Error while opening the file.\n");
      exit(EXIT_FAILURE);
   }

   printf("The contents of file are:\n");
   while((ch = fgetc(fp)) != EOF)
      printf("%c", ch);

   fclose(fp);
   return 0;
}
```

2) Compiling it as `main.wasm` using `wasm-sdk`.
3) Creating a `test.txt` with some texts inside.
4) Running it with freestanding EE:
```
RUST_LOG=info ./veracruz/sdk/freestanding-execution-engine/target/release/freestanding-execution-engine --dump-stdout true --dump-stderr true --data test.txt --program main.wasm
```

Output info:
```
[2021-08-04T07:18:21Z INFO  freestanding_execution_engine] Parsed command line.
[2021-08-04T07:18:21Z INFO  freestanding_execution_engine] Selecting interpretation as the execution strategy.
[2021-08-04T07:18:21Z INFO  freestanding_execution_engine] Using 'main.wasm' as our WASM executable.
[2021-08-04T07:18:21Z INFO  freestanding_execution_engine] Selected 1 data sources as input to computation.
[2021-08-04T07:18:21Z INFO  freestanding_execution_engine] stdout configured to be dumped before exiting.
[2021-08-04T07:18:21Z INFO  freestanding_execution_engine] stderr configured to be dumped before exiting.
[2021-08-04T07:18:21Z INFO  freestanding_execution_engine] Command line read successfully.
[2021-08-04T07:18:21Z INFO  freestanding_execution_engine] Opening file 'main.wasm' for reading.
[2021-08-04T07:18:21Z INFO  freestanding_execution_engine] WASM program main.wasm loaded into VFS.
[2021-08-04T07:18:21Z INFO  freestanding_execution_engine] Loading data source 'test.txt' with id 0 for reading.
[2021-08-04T07:18:21Z INFO  freestanding_execution_engine] Loading 'test.txt' into vfs.
[2021-08-04T07:18:21Z INFO  freestanding_execution_engine] Data sources loaded.
[2021-08-04T07:18:21Z INFO  freestanding_execution_engine] Invoking main.
[2021-08-04T07:18:21Z INFO  freestanding_execution_engine] return code: 1
[2021-08-04T07:18:21Z INFO  freestanding_execution_engine] time: 2383 micro seconds
Error while opening the file.
: Not a directory
```

**Comment by dominic-mulligan-arm** [comment]
@mofanv are the majority of these files belonging to a vendored dependency?  The changeset is 883 files in size, which makes it almost impossible to review (e.g. Github's interface slows to a crawl, for a start).  Is there some way of factoring out the changes that you've made to make this easier?

**Comment by mofanv** [comment]
> @mofanv are the majority of these files belonging to a vendored dependency? The changeset is 883 files in size, which makes it almost impossible to review (e.g. Github's interface slows to a crawl, for a start). Is there some way of factoring out the changes that you've made to make this easier?

Yes. most files are from darknet framework, although I changed the source code. I will try to pack some codes as a submodule for easy review later.

File: deep-learning-server/examples/classifier.c [hunk 1]
```diff
@@ -0,0 +1,1111 @@
...
+    int *indexes = calloc(top, sizeof(int));
+    char buff[256];
+    char *input = buff;
+    while(1){
+        if(filename){
+            strncpy(input, filename, 256);
+        }else{
+            printf("Enter Image Path: ");
+            fflush(stdout);
+            input = fgets(input, 256, stdin);
```
Comment (dominic-mulligan-arm) [inline deep-learning-server/examples/classifier.c:591]: Use more robust input mechanisms than this!

File: deep-learning-server/examples/darknet.c [hunk 2]
```diff
@@ -0,0 +1,498 @@
+#include "darknet.h"
```
Comment (dominic-mulligan-arm) [inline deep-learning-server/examples/darknet.c:1]: Is this your code @mofanv or part of the Darknet framework?  (Similar comment applies to all other files in this changeset: what is written by you, what is provided by Darknet, and what comes from Darknet but was modified by you?)

File: deep-learning-server/data/mnist/mnist_preparation.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                    help="the data distribution among clients [iid] or [non-iid]")
+parser.add_argument('--relat_path_list', type=bool, default=True,
+                    help="which to use relative image path in the loading list")
+args = parser.parse_args()
+
+
+# if MNIST file not exist, then download
+if not os.path.exists("train-images-idx3-ubyte"):
+    print("Downloading mnist files...")
+    data_files=["train-images-idx3-ubyte.gz",
```
[the remark this answers is not in the document]
Comment (mofanv) [reply deep-learning-server/data/mnist/mnist_preparation.py:31]: Currently, only downloads necessary to build the wasm binary are in the Makefile. This wasm example is more extendable and will increasingly support more applications. Thus, maybe for each use case, we put downloads, data preparation, etc in separate files could be better? Otherwise, the main makefile will be filled with things when more applications/use cases are developed.

File: deep-learning-server/Makefile [hunk 4]
```diff
@@ -0,0 +1,41 @@
+VPATH=./src/:./examples
```
[the remark this answers is not in the document]
Comment (mofanv) [reply deep-learning-server/Makefile:12]: add as below:
```
COPYRIGHT AND LICENSING

See the `LICENSE_MIT.markdown` file in the Veracruz deep learning server 
example repository root directory for copyright and licensing information.
Based on darknet, YOLO LICENSE https://github.com/pjreddie/darknet/blob/master/LICENSE
```

File: deep-learning-server/Makefile [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+EXEC=darknet.wasm
+OBJDIR=./obj/
+
+CC=/Users/fanvincentmo/Documents/wasi-sdk-12.0/bin/clang --sysroot=/Users/fanvincentmo/Documents/wasi-sdk-12.0/share/wasi-sysroot
```
[the remark this answers is not in the document]
Comment (mofanv) [reply deep-learning-server/Makefile]: the sdk is currently auto-downloaded into the parent directory. I think it will also be used by other examples built from c/c++

File: deep-learning-server/data/labels/make_labels.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import os
```
[the remark this answers is not in the document]
Comment (mofanv) [reply deep-learning-server/data/labels/make_labels.py:18]: All suitable files have been added with headers

File: deep-learning-server/data/labels/make_labels.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+import string
+import pipes
+
+font = 'futura-normal'
```
[the remark this answers is not in the document]
Comment (mofanv) [reply deep-learning-server/data/labels/make_labels.py]: removed. You're right that this is not necessary

File: deep-learning-server/data/labels/make_labels.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+# install `imagemagick` for using `convert` sh command
+def make_labels(s):
```
[the remark this answers is not in the document]
Comment (mofanv) [reply deep-learning-server/data/labels/make_labels.py:28]: Now it is detected and installed in the Makefile

File: deep-learning-server/data/mnist/mnist_preparation.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+import numpy as np
+import cv2
+import os
+import argparse
+
+num_clients = 10
+
+# arguments
+parser = argparse.ArgumentParser(description='Arguments to process the dataset')
+parser.add_argument('--num_clients', type=int, default=1,
```
[the remark this answers is not in the document]
Comment (mofanv) [reply deep-learning-server/data/mnist/mnist_preparation.py]: Done

File: deep-learning-server/data/mnist/mnist_preparation.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+                    help='the number of clients who held frictions of the dataset')
+parser.add_argument('--data_dist', type=str, default='iid',
+                    help="the data distribution among clients [iid] or [non-iid]")
+parser.add_argument('--relat_path_list', type=bool, default=True,
+                    help="which to use relative image path in the loading list")
+args = parser.parse_args()
+
+
+# if MNIST file not exist, then download
+if not os.path.exists("train-images-idx3-ubyte"):
```
[the remark this answers is not in the document]
Comment (mofanv) [reply deep-learning-server/data/mnist/mnist_preparation.py]: Currently, only downloads necessary to build the wasm binary are in the Makefile. This wasm example is more extendable and will increasingly support more applications. Thus, maybe for each use case, we put downloads, data preparation, etc in separate files could be better? Otherwise, the main makefile will be filled with things when more applications/use cases are developed.

File: deep-learning-server/data/mnist/mnist_preparation.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+else:
+    client_no = list(range(1, args.num_clients+1))
+    img_path = [cwd + "/images_client" + str(i) for i in client_no]
+    relat_img_path = ["data/mnist" + "/images_client" + str(i) for i in client_no]
+    for _path in img_path:
+        if not os.path.exists(_path):
+            os.mkdir(_path)
+
+
+# convert function
```
[the remark this answers is not in the document]
Comment (mofanv) [reply deep-learning-server/data/mnist/mnist_preparation.py]: All functions have been added with `doxygen` header, with input and output explanation

File: deep-learning-server/examples/aggregation.c [hunk 12]
```diff
@@ -0,0 +1,61 @@
+#include "darknet.h"
+
+#include <time.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+void average(int argc, char *argv[])
+{
+    char *cfgfile = argv[2];
```
[the remark this answers is not in the document]
Comment (mofanv) [reply deep-learning-server/examples/aggregation.c:38]: check added

File: deep-learning-server/examples/classifier.c [hunk 13]
```diff
@@ -0,0 +1,1111 @@
...
+    args.labels = labels;
+    if (tag){
+        args.type = TAG_DATA;
+    } else {
+        args.type = CLASSIFICATION_DATA;
+    }
+
+    data train;
+    data buffer;
+    //pthread_t load_thread;
```
[the remark this answers is not in the document]
Comment (mofanv) [reply deep-learning-server/examples/classifier.c]: removed

File: deep-learning-server/examples/classifier.c [hunk 14]
```diff
@@ -0,0 +1,1111 @@
...
+    list *options = read_data_cfg(datacfg);
+
+    char *name_list = option_find_str(options, "names", 0);
+    if(!name_list) name_list = option_find_str(options, "labels", "data/labels.list");
+    int top = option_find_int(options, "top", 1);
+
+    int i = 0;
+    char **names = get_labels(name_list);
+    //clock_t time;
+    int *indexes = calloc(top, sizeof(int));
```
[the remark this answers is not in the document]
Comment (mofanv) [reply deep-learning-server/examples/classifier.c]: All memory allocation functions have been added with fail check now

File: deep-learning-server/examples/classifier.c [hunk 15]
```diff
@@ -0,0 +1,1111 @@
...
+            printf("Enter Image Path: ");
+            fflush(stdout);
+            input = fgets(input, 256, stdin);
+            if(!input) return;
+            strtok(input, "\n");
+        }
+        image orig = load_image_color(input, 0, 0);
+        image r = resize_min(orig, 256);
+        image im = crop_image(r, (r.w - 224 - 1)/2 + 1, (r.h - 224 - 1)/2 + 1, 224, 224);
+        float mean[] = {0.48263312050943, 0.45230225481413, 0.40099074308742};
```
[the remark this answers is not in the document]
Comment (mofanv) [reply deep-learning-server/examples/classifier.c]: removed. Not used for now

File: deep-learning-server/examples/classifier.c [hunk 16]
```diff
@@ -0,0 +1,1111 @@
...
+        free_image(im);
+        if (filename) break;
+    }
+}
+
+void predict_classifier(char *datacfg, char *cfgfile, char *weightfile, char *filename, int top)
+{
+    network *net = load_network(cfgfile, weightfile, 0);
+    set_batch_network(net, 1);
+    srand(2222222);
```
[the remark this answers is not in the document]
Comment (mofanv) [reply deep-learning-server/examples/classifier.c]: removed. not necessary

File: deep-learning-server/examples/classifier.c (hunk 1 above)
Comment (mofanv) [reply deep-learning-server/examples/classifier.c]: changed to load from args file only, otherwise exit

File: deep-learning-server/examples/darknet.c (hunk 2 above)
Comment (mofanv) [reply deep-learning-server/examples/darknet.c:15]: Most codes are based darknet, and changes are made by me in order to compile it as wasm binary, and support multiple inputs, etc. Previous YOLO license + our MIT license are added to each file

File: deep-learning-server/Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (dominic-mulligan-arm) [reply deep-learning-server/Makefile]: Right, but you're trying to commit a file with a hardcoded link to your own home directory (`/Users/fanvincentmo/Documents/wasi-sdk-12.0/bin/clang`), no?

File: deep-learning-server/data/mnist/mnist_preparation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dominic-mulligan-arm) [reply deep-learning-server/data/mnist/mnist_preparation.py:38]: Not sure you replied to the right comment, or misunderstood what I was asking: I meant hoisting out the names of the Gzip files into top-level constants in the Python file.

File: deep-learning-server/Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (mofanv) [reply deep-learning-server/Makefile]: Sorry, this is a misunderstanding. What I meant is that after changing, it is: 
```
CC=../wasi-sdk-12.0/bin/clang --sysroot=../wasi-sdk-12.0/share/wasi-sysroot
```
`make` target will do:
```
wasisdk:
	if [ ! -d "../wasi-sdk-12.0" ]; then \
		wget -P ../ https://github.com/WebAssembly/wasi-sdk/releases/download/wasi-sdk-12/wasi-sdk-12.0-$(OS).tar.gz; \
		mkdir ../wasi-sdk-12.0; \
		tar xvf ../wasi-sdk-12.0-$(OS).tar.gz -C ../; \
		rm -f ../wasi-sdk-12.0-macos.tar.gz; \
	fi
```

`wasi-sdk-12.0` will exist in the parent directory, so other examples can use it as well

[the remark this answers is not in the document]
Comment (dominic-mulligan-arm) [reply deep-learning-server/Makefile]: Ah OK, that makes sense.

File: deep-learning-server/data/mnist/mnist_preparation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mofanv) [reply deep-learning-server/data/mnist/mnist_preparation.py]: Misunderstanding here. I thought of putting them in the makefile.
Updated. Filenames are constants in the python file now.

[the remark this answers is not in the document]
Comment (mofanv) [reply deep-learning-server/data/mnist/mnist_preparation.py]: See lines starting from: https://github.com/mofanv/veracruz-examples/blob/main/deep-learning-server/data/mnist/mnist_preparation.py#L34

File: deep-learning-server/README.md [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+To test on one image, using the commands below. You will see the prediction of this image.
+```
+cp args_files/prediction_one.cfg args.cfg
+wasmtime --dir=./ dl-server.wasm
+
+# or using freestanding execution engine
+RUST_LOG=“info” [FREESTANDING EE EXECUTABLE] --program dl-server.wasm --input-source args.cfg cfg model data/mnist/mnist.names.list data/mnist/images_client1/t_00000_c5.png -d -e
+```
+
+To test on the validation dataset (10000 images), using the commands below. You will see all predictions and the top 1 and top 5 average accuracies.
```
Comment (dominic-mulligan-arm) [inline deep-learning-server/README.md:47]: sp: using --> use

File: deep-learning-server/data/labels/make_labels.py [hunk 18]
```diff
@@ -0,0 +1,44 @@
+# This file uses `convert` system function to convert character (such as
```
Comment (dominic-mulligan-arm) [inline deep-learning-server/data/labels/make_labels.py:1]: sp: uses `convert` --> uses the `convert`

File: deep-learning-server/examples/aggregation.c [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+            if (dt == 1 || dt == 14){
+                int ts = model_sum->graph->initializer[j]->n_float_data;
+                float *td_sum = model_sum->graph->initializer[j]->float_data;
+                float *td_c = model_c->graph->initializer[j]->float_data;
+                for(int k=0; k < ts; k++){
+                    td_sum[k] = td_sum[k] + td_c[k];
+                }
+                
+            // INT32, INT16, INT8, UINT16, UINT8, BOOL, or FLOAT16
+            }else if (dt==2 || dt==3 || dt==4 || dt==5 || dt==6 || dt==9 || dt==10){
```
Comment (dominic-mulligan-arm) [inline deep-learning-server/examples/aggregation.c:137]: Bizarre spacing around `}else` here and elsewhere in this file.

File: deep-learning-server/examples/darknet.c [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+    char *args_file = "args.cfg";
+    FILE *fp = fopen(args_file, "r");
+
+    // if number of args >= 2 (the first is the name of wasm itself),
+    // run example based on cmd line args directly
+    if(argc >= 2){
+        run_example(argc, argv);
+
+    // else if the args file exists, read files (and run example)
+    }else if(fp != NULL){
```
Comment (dominic-mulligan-arm) [inline deep-learning-server/examples/darknet.c:109]: Spacing issues around `else` again, here.

File: deep-learning-server/examples/detector.c [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+    char buff[256];
+    char *input = buff;
+    float nms=.45;
+
+    if(!filename){
+        fprintf(stderr, "file not exists: %s\n", filename);
+
+    }else{
+        // load image
+        strncpy(input, filename, 256);
```
Comment (dominic-mulligan-arm) [inline deep-learning-server/examples/detector.c:48]: Why is this needed?  Can't you just pass `filename` to `load_image_color`?

File: deep-learning-server/examples/aggregation.c (hunk 19 above)
Comment (mofanv) [reply deep-learning-server/examples/aggregation.c:139]: all of these are formatted

File: deep-learning-server/examples/detector.c (hunk 21 above)
Comment (mofanv) [reply deep-learning-server/examples/detector.c:49]: this `input` buff was considered for multiple images in streaming. For example, when a video is decoded as many images, the detector works for a bunch of images. I'm removing the buff, as it is not necessary for now.

**Review by ShaleXIONG — APPROVED** [review]
